### PR TITLE
Miscellaneous performance & QOL changes

### PR DESCRIPTION
**Pull request by dgant**
I saw some activity on master and I know there's interest in releasing a 2.0 at some point, so thought I'd post some of my miscellaneous local changes if folks would like them. 

There are three kinds of changes in here:
- Performance optimization: Allowing more access to BWAPI data without having to create temporary objects. These objects are free in the C++ API because they just live on the stack, but in Java cause heap allocation which makes their cost non-trivial when called frequently.
- Giving access to the IDs of some enumerated types. This allows storing them in an array and doing ID->type lookups, which I found handy
- Exposing the underlying BWEM object through BWTA. This allows you to use the BWTA interface in your bot but pass the BWEM object to, say, JBWEB

Happy to separate or remove any of these if there isn't agreement on their inclusion.

### CODE REVIEW

File: src/main/java/bwapi/Game.java [hunk 1]
```diff
@@ -872,10 +872,7 @@ final public boolean isWalkable(final int walkX, final int walkY) {
     }
 
     final public boolean isWalkable(final WalkPosition position) {
-        if (!position.isValid(this)) {
-            return false;
-        }
-        return walkable[position.x][position.y];
+        return position.isValid(this) && walkable[position.x][position.y];
```
Comment (dgant) [inline src/main/java/bwapi/Game.java:875]: Purely cosmetic syntax simplification.

File: src/main/java/bwapi/Game.java [hunk 2]
```diff
@@ -892,15 +889,12 @@ final public boolean isWalkable(final WalkPosition position) {
...
+        return isValidTile(tileX, tileY) ? groundHeight[tileX][tileY] : 0;
     }
 
-    public int getGroundHeight(final TilePosition position) {
-        if (!position.isValid(this)) {
-            return 0;
-        }
-        return groundHeight[position.x][position.y];
+    final public int getGroundHeight(final TilePosition position) {
+        return getGroundHeight(position.x, position.y);
```
Comment (dgant) [inline src/main/java/bwapi/Game.java:897]: Allowing access to `getGroundHeight()` without having to allocate a TilePosition.

File: src/main/java/bwapi/Game.java [hunk 3]
```diff
@@ -1014,23 +1008,23 @@ final public boolean hasPowerPrecise(final Position position, final UnitType uni
...
     }
 
-    final public boolean hasPower(final TilePosition position, final UnitType unitType) {
+    final public boolean hasPower(final int tileX, final int tileY, final UnitType unitType) {
         if (unitType.id >= 0 && unitType.id < UnitType.None.id) {
-            return hasPowerPrecise(position.x * 32 + unitType.tileWidth() * 16, position.y * 32 + unitType.tileHeight() * 16, unitType);
+            return hasPowerPrecise(tileX * 32 + unitType.tileWidth() * 16, tileY * 32 + unitType.tileHeight() * 16, unitType);
         }
-        return hasPowerPrecise(position.x * 32, position.y * 32, UnitType.None);
+        return hasPowerPrecise(tileY * 32, tileY * 32, UnitType.None);
```
Comment (dgant) [inline src/main/java/bwapi/Game.java:1027]: Allowing access to `hasPower[Precise]()` without having to allocate a TilePosition.

File: src/main/java/bwapi/Unit.java [hunk 4]
```diff
@@ -835,14 +835,17 @@ public UnitType getBuildType() {
...
+
+    public UnitType getTrainingQueueAt(int i) {
+        if (game.isLatComEnabled() && self().trainingQueue[i].valid(game.getFrameCount())) {
+            return self().trainingQueue[i].get();
+        }
+        return UnitType.idToEnum[unitData.getTrainingQueue(i)];
     }
 
-    int getTrainingQueueCount() {
+    public int getTrainingQueueCount() {
```
Comment (dgant) [inline src/main/java/bwapi/Unit.java:860]: Two new methods `getTrainingQueueAt()` and `getTrainingQueueCount()` allow peeking at the training queue without creating a new collection each time.

Comment (JasperGeurtz) [reply src/main/java/bwapi/Unit.java:860]: I like it, can you add javadoc similar to the other methods? I think we should also mention in those docs that they are not exposed by native BWAPI

File: src/main/java/bwapi/UnitType.java [hunk 5]
```diff
@@ -252,7 +252,7 @@ public enum UnitType {
         Arrays.stream(UnitType.values()).forEach(v -> idToEnum[v.id] = v);
     }
 
-    final int id;
+    public final int id;
```
Comment (JasperGeurtz) [inline src/main/java/bwapi/UnitType.java:255]: If we expose `id`, I think it's best to add it to all Enums for consistency reasons:
 - Race
 - BulletType
 - CommandType
 - CoordinateType
 - EventType
 - GameType
 - Latency
 - Order
 - PlayerType
 - ShapeType
 - TechType
 - UnitType
 - UpgradeType

**Comment by JasperGeurtz** [comment]
LGTM

**Pull request merged**